content/post/mnist-example.md

```diff
@@ -0,0 +1,148 @@
+
+
+
+## Traning vs Inference
+
+Before we begin looking at any code, we need to make a small distinction.  A machine learning process has two distinct stages: training and inference.  During the training phase, we are creating the model itself, and we prepare it using a dataset that has been designed to do so. Typically, one would then take the trained model and use it repeatedly to make inferences about new, unseen data (hence the "inference" name for the phase).
+
+Whilst training is indeed a fundamental part of the machine learning process, stream computing lends itself much better to those situations where the model is being used for inference, perhaps as part of a bigger pipeline which may include data pre-processing and result intrepretation. Therefore, this blog entry concentrates on how we can use an existing trained model to make inferences on a very data stream of images.
+
+## Creating the models
+
+Even though we have said that the focus will be on inference, we still need to create some models to be able to use them. In this case, all the code for training is contained in `train_models.py`. We invite you to take a detailed look at it, but for the sake of this blog entry we only need to know that it is training a PCA for data preprocessing and a logistic regression for classification. The two models are then serialized to disk using sklearn's built-in pickle compatibility, to two separate file: `pca.model` and `logistic.model`.
+
+## Application Setup
+
+Our application will contain two separate computations, one for the PCA transformation and one for the classification itself.
+We set up our Wallaroo application as follows
+
+```python
+def application_setup(args):
+    with open('pca.model', 'r') as f:
+        pca = pickle.load(f)
+    with open('logistic.model', 'r') as f:
+        logistic = pickle.load(f)
+    global pca
+    global logistic
+
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("MNIST classification")
+    ab.new_pipeline("MNIST",
+                    wallaroo.TCPSourceConfig(in_host, in_port, decode))
+    ab.to(pca_transform)
+    ab.to(logistic_classify)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, encode))
+    return ab.build()
+```
+
+Note how we are loading our model in `application_setup` and making them global.  This is because loading a pickled scikit-learn model could be a potentially expensive operation, and we want to ensure we do it only once per worker. By doing so in the application_setup, the models are only loaded during initialization, and made available to the computations via the `global` keyword.
```

Comment:
This is good but I think you should explain what the flow of data that the application builder is setting up,